substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/oldobject/TestOldObjectProfiler.java

```diff
@@ -0,0 +1,384 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.test.jfr.oldobject;
+
+import com.oracle.svm.core.Uninterruptible;
+import com.oracle.svm.core.jfr.oldobject.OldObjectEffects;
+import com.oracle.svm.core.jfr.oldobject.OldObjectProfiler;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.lang.ref.WeakReference;
+import java.util.Arrays;
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+public class TestOldObjectProfiler {
+    @Test
+    public void testDoNotEmitEventsNewerThanLastSweep() {
+        final int size = 8;
+        final TestEffects effects = new TestEffects(20L, size);
+        final OldObjectProfiler profiler = new OldObjectProfiler(size, effects);
+        for (int i = 0; i < size; i++) {
+            profiler.sample(new WeakReference<>(String.valueOf(i)), i * 100, -1);
+        }
+
+        profiler.emit(0, 10);
+        Assert.assertEquals(0, effects.sizeSamples());
+    }
+
+    @Test
+    public void testSampleAfterEmit() {
+        final int size = 8;
+        final TestEffects effects = new TestEffects(20L, size) {
+            @Override
+            @Uninterruptible(reason = "Accesses allocation profiler.")
+            public boolean isAlive(WeakReference<?> ref) {
+                if (ticks <= 29) {
+                    return true;
+                }
+                final Object obj = ref.get();
+                return obj != null && ((int) ref.get()) >= 11;
+            }
+        };
+        final OldObjectProfiler profiler = new OldObjectProfiler(size, effects);
+
+        for (int i = 0; i < size; i++) {
+            profiler.sample(new WeakReference<>(i), i * 100, -1);
+        }
+
+        profiler.emit(0, Long.MAX_VALUE);
+        assertStreamEquals(IntStream.rangeClosed(0, 7).boxed(), effects.objects());
+
+        for (int i = 1; i <= size / 2; i++) {
+            profiler.sample(new WeakReference<>(10 + i), i * 1000, -1);
+        }
+
+        // Clear accumulated samples and see what gets emitted now.
+        effects.clearSamples();
+        profiler.emit(0, Long.MAX_VALUE);
+        assertStreamEquals(IntStream.rangeClosed(11, 14).boxed(), effects.objects());
+    }
+
+    @Test
+    public void testScavengeMiddle() {
+        final int size = 8;
+        final TestEffects effects = new TestEffects(20L, size) {
+            @Override
+            @Uninterruptible(reason = "Accesses allocation profiler.")
+            public boolean isAlive(WeakReference<?> ref) {
+                return !"4".equals(ref.get());
+            }
+        };
+        final OldObjectProfiler profiler = new OldObjectProfiler(size, effects);
+
+        for (int i = 0; i < size; i++) {
+            profiler.sample(new WeakReference<>(String.valueOf(i)), i * 100, -1);
+        }
+
+        profiler.emit(0, Long.MAX_VALUE);
```

Comment:
Hmmmm, I think I know what you mean, but let's refresh how scavenging has been ported to see if this makes sense:

Technically the way `scavenge` works in HotSpot is like this:

```
void ObjectSampler::add(HeapWord* obj, size_t allocated, traceid thread_id, bool virtual_thread, const JfrBlobHandle& bh, JavaThread* thread) {
  ...

  if (Atomic::load(&_dead_samples)) {
    // There's a small race where a GC scan might reset this to true, potentially
    // causing a back-to-back scavenge.
    Atomic::store(&_dead_samples, false);
    scavenge();
  }
```

It only gets called when sampled, but it calls it whenever a sample has been GC'd. There's no such notification mechanism in SVM, so instead I went for the option of scavenging when the queue is full.

Now, the issue I think you're implying above is that if the queue is not full, no scavenging happens and maybe it's good idea to scavenge after emitting to see if anything is dead and can be removed. This sounds like a good idea to me. It seems better than an alternative option of scavenging in each object sample add call. Have I got this right?

Reply:
> You may not even need to run a full `scavenge()` operation. When emitting in `OldObjectEventEmitter.emitUnchained(long)` you already check if each sample is alive. If it is dead, you can directly remove it with `OldObjectSampler.remove(OldObject)`. Admittedly, this is only really useful if you have multiple simultaneous recordings running since we only emit the OldObjectSample event once per recording anyway (so usually there will be no point to doing the extra cleanup work during event emission).

Ok, I think I'll leave the logic as is for now.

> hmm my original comment was actually concerning the test itself. I think that your current implementation of scavenging/emitting works perfectly fine. I think `testScavengeMiddle`, `testScavengeYoungest`, `testScavengeOldest` and `testScavengeAll` don't actually exercise the code in `OldObjectSampler.scavenge()` and instead rely on `OldObjectEventEmitter.emitUnchained(long)` to remove the dead samples.

That's right. I'll rework this to make sure the scavenging gets tested.